eth_tester/main.py

```diff
@@ -64,6 +69,28 @@ def get_default_fork_blocks():
     }
 
 
+def handle_auto_mining(func):
+    @functools.wraps(func)
+    def func_wrapper(self, *args, **kwargs):
+        if not self.auto_mine_transactions:
+            snapshot = self.take_snapshot()
+
+        transaction_hash = func(self, *args, **kwargs)
+
+        if self.auto_mine_transactions:
+            self.mine_block()
+        else:
+            pending_transaction = self.get_transaction_by_hash(transaction_hash)
+            # Remove any pending transactions with the same nonce
+            self._pending_transactions = remove_matching_transaction_from_list(
+                self._pending_transactions, pending_transaction)
+            self._pending_transactions.append(pending_transaction)
+            self.revert_to_snapshot(snapshot)
```

Comment:
I'm thinking this `revert` might need to be in a try/finally block. Otherwise, if there is any kind of exception after a state change, we will fail to roll back.

Reply:
I'm not sure I understand. Isn't `revert_to_snapshot` doing the rollback? And if that fails, is there really a meaningful way to continue? 

It feels to me like the whole function should completely cleanly otherwise the final state is not really usable

Reply:
Sorry, what I meant was something like:

```py
snapshot = self.take_snapshot()

try:
    transaction_hash = func(self, *args, **kwargs)
    pending_transaction = self.get_transaction_by_hash(transaction_hash)
    ...
    self._pending_transactions.append(pending_transaction)
finally:
    self.revert_to_snapshot(snapshot)
```
```

Reply:
Oh 😄 I see what you mean. That's it updated.